Decomposition of bmm, addmm, mm for dot product

mikekgfb opened the pull request:
Summary: Decomposition of bmm, addmm, mm for dot product 

Test Plan:
sandcastle ; github ci

Differential Revision: D47881983



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

jansel commented:
Looks like there are some stray prints.  Any perf numbers?

mikekgfb commented:
@EikanWang @jgong5 I've found that during code generation with GCC, it mostly uses XMM 128b registers to generate code, rather than ymm 256b registers, even when setting -fpreferred-vector-width=256.  Does this represent an issue with GCC vectorization, or is the distribution of function units such that xmm register use is preferable for most compute-bound applications?

mikekgfb commented:
> Looks like there are some stray prints. Any perf numbers?

+ 50% or more on workloads measured, for CPU code.

jgong5 commented:
> @EikanWang @jgong5 I've found that during code generation with GCC, it mostly uses XMM 128b registers to generate code, rather than ymm 256b registers, even when setting -fpreferred-vector-width=256. Does this represent an issue with GCC vectorization, or is the distribution of function units such that xmm register use is preferable for most compute-bound applications?

What machine arch did you use? Can you share the generated code? I can take a look.

Commit:
Decomposition of bmm, addmm, mm for dot product (#106219)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106219

Decomposition of bmm, addmm, mm for dot product

Test Plan: sandcastle ; github ci

Differential Revision: D47881983

fbshipit-source-id: 50fb5a6e7ef255eb12da596498bd41058f42b892

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -177,6 +177,30 @@ def baddbmm(self, batch1, batch2, beta=1, alpha=1):
+    return NotImplemented
+
+@register_decomposition([aten.addmm])
+def addmm(self, input1, input2, beta=1, alpha=1):
```
Is there any problem with the default decomposition at https://github.com/pytorch/pytorch/blob/3d165dc3f3e1ba42c7512200b1b9c14d4556ec61/torch/_decomp/decompositions.py#L1177 so that you have to define it again here?

mikekgfb replied:
Good point - this code went through several iterations. The initial code had both the addmm expansion and dot product and skinny matrix recognition in one function.  After a refactor, addmm now mirrors the code you pointed to.  I might test the present code without the addmm rule

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -177,6 +177,30 @@ def baddbmm(self, batch1, batch2, beta=1, alpha=1):
+
+@register_decomposition([aten.mm])
+def mm(self, input2):
+    if self.size(-1) == 1 and input2.size(0) == 1 and (self.dtype == input2.dtype) and ((torch.numel(self)+torch.numel(input2)) <= 32):
```
How is this threshold decided? According to the performance numbers?

jgong5 reviewed (commented):
Can you share detailed performance comparison before and after the PR on various shapes in concern? Also, do we have UTs covering these shapes?

mikekgfb commented:
That is the goal.

There are several steps still to take:
1 - understand how well this applies to different targets. For example a platform that can support adding with alpha and beta factors in addition to the individual components may do better calling the kernel than decomposing the compression. Also if it can merge an activation function into this kernel “for free”

2 - establish what the special optimization for [k, n ] * [n, m] should be to take advantage of skinny n tensor. And what the limit is on the size - this is where I use the size cutoff right now. It helps definitely for small values, but how much? Does it change with generations of hardware in addition to hardware sku? (The current check is also defective as inumel does not consider broadcasted dimensions in the current form.)

3 - how much loop merge and libraries optimization does this enable? W across for product and loop jam etc we were able to achieve great performance. We might need input from platform experts like yourself to establish values that work across broad range of skus. (Maybe a hardware profile for a small set of choices like mobile vs serve devices )

The question appears to be solvable with with Autotuning but that is too expensive to make
For so many combinatorial choices. We know that dot product win [1, n] * [n, 1] delivers huge end to end wins.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Jiong Gong ***@***.***>
Sent: Thursday, August 3, 2023 3:44:10 AM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] Decomposition of bmm, addmm, mm for dot product (PR #106219)

@jgong5 commented on this pull request. Can you share detailed performance comparison before and after the PR on various shapes in concern? Also, do we have UTs covering these shapes? In torch/_inductor/decomposition. py: > @@ -177,6 +177,30
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

@jgong5 commented on this pull request.

Can you share detailed performance comparison before and after the PR on various shapes in concern? Also, do we have UTs covering these shapes?

________________________________

In torch/_inductor/decomposition.py<https://github.com/pytorch/pytorch/pull/106219#discussion_r1282777378>:

> @@ -177,6 +177,30 @@ def baddbmm(self, batch1, batch2, beta=1, alpha=1):
         self = self * beta
     return self + result

***@***.***_decomposition([aten.bmm])
+def bmm(self, batch2):
+    if self.size(1) == 1 and batch2.size(-1) == 1:
+        return torch.sum(self.squeeze(1) * batch2.squeeze(-1), dim=1, keepdim=True).unsqueeze(1)
+    return NotImplemented
+
***@***.***_decomposition([aten.addmm])
+def addmm(self, input1, input2, beta=1, alpha=1):


Is there any problem with the default decomposition at https://github.com/pytorch/pytorch/blob/3d165dc3f3e1ba42c7512200b1b9c14d4556ec61/torch/_decomp/decompositions.py#L1177 so that you have to define it again here?

________________________________

In torch/_inductor/decomposition.py<https://github.com/pytorch/pytorch/pull/106219#discussion_r1282784743>:

> +    return NotImplemented
+
***@***.***_decomposition([aten.addmm])
+def addmm(self, input1, input2, beta=1, alpha=1):
+    result = torch.mm(input1, input2)
+    if not isinstance(alpha, numbers.Number) or alpha != 1:
+        result = result * alpha
+    if beta == 0:
+        return result
+    if not isinstance(beta, numbers.Number) or beta != 1:
+        self = self * beta
+    return self + result
+
***@***.***_decomposition([aten.mm])
+def mm(self, input2):
+    if self.size(-1) == 1 and input2.size(0) == 1 and (self.dtype == input2.dtype) and ((torch.numel(self)+torch.numel(input2)) <= 32):


How is this threshold decided? According to the performance numbers?

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/106219#pullrequestreview-1560427003>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHIXBMJSP6Y4XAOFI4TXTNJEVANCNFSM6AAAAAA235CQRE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

jgong5 commented:
> That is the goal.

Still want to clarify your target. Are you going to use codegen based approach to replace the library-based matmuls or just focus on some degenerated matmul cases, e.g., inner/outer product of two vectors? I see two reasons why the codegen based approach might be better than library: 1) save of library overhead from compute with small shapes; 2) additional op fusions that the library doesn't cover (note that the library already covers some of the post-op fusions), and still this is with small enough shapes. Otherwise, the library-based approach should be efficient enough.

To tackle the codegen of GEMM in general, we usually apply a hybrid approach - loopnests plus micro-kernels, e.g., the brgemm (https://www.frontiersin.org/articles/10.3389/fams.2022.826269/full) abstraction in oneDNN. The loopnests take care of the thread decomposition and cache blocking while the micro-kernels on all the HW arch specific things (e.g., register blocking, loop unrolling, data prefetching, instruction selection and scheduling etc.). Different loop schedules and micro-kernel options would be needed for different shapes to gain the peak performance.

If you are only interested in the degenerated cases, Pure codegen based approach should be fine (as what you already implemented). But I suppose these are computation with trivial shapes and tiny ops by themselves? I'm curious what model you were running that saw huge e2e wins.

mikekgfb commented:
> > That is the goal.
> 
> Still want to clarify your target. Are you going to use codegen based approach to replace the library-based matmuls or just focus on some degenerated matmul cases, e.g., inner/outer product of two vectors? I see two reasons why the codegen based approach might be better than library: 1) save of library overhead from compute with small shapes; 2) additional op fusions that the library doesn't cover (note that the library already covers some of the post-op fusions), and still this is with small enough shapes. Otherwise, the library-based approach should be efficient enough.

My immediate goal is to take care of degenerate cases that suffer in particular on small inputs (we're talking 40-element dot products etc.  These are exacerbated by the cost of library calls, and the inability to do loop fusion, and pointwise optimizations

> To tackle the codegen of GEMM in general, we usually apply a hybrid approach - loopnests plus micro-kernels, e.g., the brgemm (https://www.frontiersin.org/articles/10.3389/fams.2022.826269/full) abstraction in oneDNN. The loopnests take care of the thread decomposition and cache blocking while the micro-kernels on all the HW arch specific things (e.g., register blocking, loop unrolling, data prefetching, instruction selection and scheduling etc.). Different loop schedules and micro-kernel options would be needed for different shapes to gain the peak performance.

We'll need to tackle more general cases for performance, and the hybrid approach is a very viable solution - at least for some (server?) platforms.  For other targets (e.g., embedded/mobile targets), we may have different optimization criteria, such as code size and calling a library for all but the most degenerate cases may be a better solution.

> If you are only interested in the degenerated cases, Pure codegen based approach should be fine (as what you already implemented). But I suppose these are computation with trivial shapes and tiny ops by themselves? I'm curious what model you were running that saw huge e2e wins.

As I mention above, we saw some pretty bad performance on degenerate cases that needed an immediate fix.

There's a cut-down submission of this PR in https://github.com/pytorch/pytorch/pull/106593 that focuses on these degenerate cases with the goal of landing quickly for a fix for CPU code gen in those cases while we design a solution for more general tensor multiplies, larger inputs, and other platforms.

jansel reviewed (approved):
Would be good to have tests for this.

mikekgfb commented:
A version  of these transformations applied to cpu- resident tensors only has landed, including unit tests.

For other targets, we’ll want to find a way to determine when and which of these transforms make sense - this might eventually need us to introduce some sort of cost function for different targets.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Jason Ansel ***@***.***>
Sent: Saturday, August 26, 2023 6:18:01 AM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] Decomposition of bmm, addmm, mm for dot product (PR #106219)

@jansel approved this pull request. Would be good to have tests for this. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread. Message ID: <pytorch/pytorch/pull/106219/review/1596281090@ github. com>
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

@jansel approved this pull request.

Would be good to have tests for this.

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/106219#pullrequestreview-1596281090>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHINXY3GEW4VCMTPBG3XXDT5TANCNFSM6AAAAAA235CQRE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
